src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2247,6 +2247,42 @@ internal void AddModuleInitializerMethod(MethodSymbol method)
             LazyInitializer.EnsureInitialized(ref _moduleInitializerMethods).Add(method);
         }
 
+        private ConcurrentSet<(InterceptsLocationAttributeData, MethodSymbol)>? _interceptions;
+
+        internal void AddInterception(InterceptsLocationAttributeData location, MethodSymbol interceptor)
+        {
+            Debug.Assert(!_declarationDiagnosticsFrozen);
+            LazyInitializer.EnsureInitialized(ref _interceptions).Add((location, interceptor));
+        }
+
+        internal (InterceptsLocationAttributeData data, MethodSymbol interceptor)? GetInterceptor(Location? callLocation)
+        {
+            if (_interceptions is null || callLocation is null)
+            {
+                return null;
+            }
+
+            foreach (var (interceptsLocation, interceptor) in _interceptions)
```

Comment:
>foreach (var (interceptsLocation, interceptor) in _interceptions) [](http://example.com/codeflow?start=12&length=65)

We should test this with a large number of intercepted calls.

Reply:
Do agree, but would like to punt it to a future PR.